File: 853-car-fleet/853-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        # if one ahead arrives later or at same time, they intersect
        # if fleets join, can "pretend" theyre one to make later 
        # calculations easier
        
        #if fleet intersects, then poistion is min, and the slope is
        # (start - end)/ (endTime)
        # endTime is still the same as previous fleet
        
        def intersect(x, y):
            # if y gets there later they hit
            # x[0] + x[1]*t = target
            # t = (target - x[1]) /x[0]
            xTime = (target - x[0]) / x[1]
            yTime = (target - y[0]) / y[1]
            if xTime <= yTime:
                return True, yTime
            return False, xTime
        
        posSort = sorted(zip(position, speed), key = lambda x: target - x[0])
        stack = []
        
        for p, s in posSort:
            if stack:
                intersec, time = intersect((p,s), stack[-1])
                if not intersec:
                    stack.append([p, s])
                    continue
                stack[-1] = [p, (target - p)/time]
            else:
                stack.append([p, s])
        # print(stack, p,s)
        return len(stack)

File: 853-car-fleet/853-car-fleet.cpp
class Solution {
public:
    
    static bool comparePos(tuple<int,int> i1, tuple<int,int> i2)
    {
        return (get<0>(i1) > get<0>(i2));
    }
    
    double destTime(const int target, const tuple<int, int>& car){
        return double(target - get<0>(car)) / double(get<1>(car));
    }
  
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<tuple<int,int>>sp(0);
        int ans = 0;
        
        for(int i = 0; i < speed.size(); ++i){
            sp.push_back({position[i], speed[i]});
        }
        
        sort(sp.begin(), sp.end(), Solution::comparePos);
        
        double latestTime = destTime(target, sp[0]);
        int count = 1;
        
        for(auto& car: sp){
            double arrival = destTime(target, car);
            // cout<< arrival << " " << stack.back() << "\n";
            if (arrival > latestTime){
                latestTime = arrival;
                count++;
            }
        }
        
        return count;
    }
};